### PR TITLE
Add new crate for ics101 integration

### PR DESCRIPTION
**Pull request by luckiester**
Add new crate for ics101 integration

### CODE REVIEW

File: contracts/ics101/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+            .div(Uint128::from(remote_asset_amount_in_pool.balance.amount));
+
+        check_slippage(ratio_of_tokens_in, ratio_of_assets_in_pool, 10)
+            .map_err(|err| StdError::generic_err(format!("Invalid Slippage: {}", err)))?;
+    }
+
+    // Create the interchain market maker
+    let fee = MAX_FEE_RATE;
+    // let amm = InterchainMarketMaker::new(pool.clone(), fee);
+    let amm = InterchainMarketMaker {
```
Comment (liangping) [inline contracts/ics101/src/contract.rs:202]: pool and pool id are duplicated arguemnts

File: contracts/ics101/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+}
+
+fn singe_asset_withdraw(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: MsgSingleAssetWithdrawRequest,
+) -> Result<Response, ContractError> {
+    // Get liquidity pool
+    let pool = POOLS.load(deps.storage, &msg.pool_coin.denom)?;
```
Comment (liangping) [inline contracts/ics101/src/contract.rs:248]: Do we need check pool status here.

File: contracts/ics101/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    let amm = InterchainMarketMaker::new(&pool, fee);
+
+    let local_out = amm.multi_asset_withdraw(
+        msg.local_withdraw.pool_coin.clone(),
+        &msg.local_withdraw.denom_out,
+    )?;
+    let remote_out = amm.multi_asset_withdraw(
+        msg.remote_withdraw.pool_coin.clone(),
+        &msg.remote_withdraw.denom_out,
+    )?;
```
Comment (liangping) [inline contracts/ics101/src/contract.rs:307]: How to make sure that the sender has enough pool tokens on remote chain?

File: contracts/ics101/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        });
+    }
+
+    let fee = MAX_FEE_RATE;
+    let amm = InterchainMarketMaker::new(&pool, fee);
+
+    let local_out = amm.multi_asset_withdraw(
+        msg.local_withdraw.pool_coin.clone(),
+        &msg.local_withdraw.denom_out,
+    )?;
```
Comment (liangping) [inline contracts/ics101/src/contract.rs:303]: need to check if the sender have enough pool tokens

File: contracts/ics101/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+    // Construct the IBC data packet
+    let swap_data = to_binary(&msg)?;
+
+    let fee = MAX_FEE_RATE;
+    let amm = InterchainMarketMaker::new(&pool, fee);
+
+    let token_out: Option<Coin>;
+    let msg_type: Option<SwapMessageType>;
+
+    match msg.swap_type {
```
Comment (liangping) [inline contracts/ics101/src/contract.rs:363]: need check if the sender has deposited in tokens

Comment (liangping) [reply contracts/ics101/src/contract.rs:363]: and also need check deposit amount should equal msg.token_in

File: contracts/ics101/src/market.rs [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        if self.pool.status == PoolStatus::PoolStatusInitial {
+            let total_initial_lp_amount: Uint128 = self
+                .pool
+                .assets
+                .iter()
+                .map(|a| a.balance.amount.clone())
+                .sum();
+
+            issue_amount = total_initial_lp_amount;
+        } else {
```
Comment (liangping) [inline contracts/ics101/src/market.rs:118]: we can remove this now. pool will become `ready` status once it's created.

**Review by liangping — CHANGES_REQUESTED** [review]

**Comment by amityadav0** [comment]
Merged via different branch #9 #10

**Pull request closed without merging**